crates/app/src/builder.rs

```diff
@@ -3,32 +3,63 @@
 
 use super::{handle, RepoStore, TagStore, TreeStore};
 
+use drawbridge_store::Storage;
+
+use std::error::Error;
+use std::ffi::OsStr;
+use std::fs::{create_dir, File};
+use std::path::Path;
 use std::sync::Arc;
 
 use axum::handler::Handler;
 use axum::routing::IntoMakeService;
 use axum::{Extension, Router};
+use cap_async_std::fs::Dir;
+use tokio::sync::RwLock;
+
+pub struct Builder<'a> {
+    data_directory: &'a Path,
+}
 
-#[derive(Default)]
-pub struct Builder;
+impl<'a> Default for Builder<'a> {
+    fn default() -> Self {
+        Self {
+            data_directory: Path::new("drawbridge-data"),
+        }
+    }
+}
 
-impl Builder {
+impl<'a> Builder<'a> {
+    /// Constructs a new [Builder].
     pub fn new() -> Self {
-        Self
+        Default::default()
     }
 
-    // TODO: Add configuration functionality
+    /// The directory to be used for all data storage and configuration.
+    pub fn data_directory<S: AsRef<OsStr> + ?Sized>(mut self, path: &'a S) -> Self {
+        self.data_directory = Path::new(path);
+        self
+    }
 
     /// Builds the application and returns Drawbridge instance as a [tower::MakeService].
-    pub fn build(self) -> IntoMakeService<Router> {
-        let repos: Arc<RepoStore> = Default::default();
-        let tags: Arc<TagStore> = Default::default();
-        let trees: Arc<TreeStore> = Default::default();
-        Router::new()
+    pub fn build(self) -> Result<IntoMakeService<Router>, Box<dyn Error>> {
+        let path = Path::new(&self.data_directory);
+
+        if !path.exists() {
+            create_dir(path)?;
+        }
+
+        let root = Dir::from_std_file(File::open(path)?.into());
+        let repos: Arc<RepoStore> =
+            Arc::new(RwLock::new(Storage::new(root.clone(), "repo".to_string())));
```

Comment:
The locking should happen on filesystem level, not in-memory (because there might be `N` drawbridge instances running at the same time), we can file a follow-up issue about on-disk locking, but we need to account for that in the design, *if it is required at all*. If possible, we should not require any locking at all in the filesystem storage design.

If we do require a lock (preferably, we should not), then the mutex should be moved into the store implementation.